### PR TITLE
move obo call to earlier

### PR DESCRIPTION
**Pull request by jennyf19**
#381 
cc: @jmprieur 
Needs a lot of manual testing:
- [x] WebAppCallsWebApiCallsGraph
- [x] B2C WebAppCallsWebApi

### CODE REVIEW

**Comment by jmprieur** [comment]
@jennyf19 tested the Web API incremental tutorial which is now way faster when doing OBO. This is perceivable from the UI of the client when adding a todo item it's now super fast.

File: src/Microsoft.Identity.Web/Constants/LogMessages.cs [hunk 1]
```diff
@@ -21,5 +21,7 @@ internal static class LogMessages
         public const string SessionCacheKeyNotFound = "Cache key {0} not found in session {1}. ";
         public const string SerializingSessionCache = "Serializing session {0}, cache key {1}. ";
         public const string ClearingSessionCache = "Clearing session {0}, cache key {1}. ";
+
+        public const string ErrorAcquiringTokenForOboForWebApi = "Error acquiring a token for obo for a web API - MsalUiRequiredException message is: {0} .";
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/Constants/LogMessages.cs:25]:
```suggestion
        public const string ErrorAcquiringTokenForOboForWebApi = "Error acquiring a token for On-Behalf-Of flow for a web API - MsalUiRequiredException message is: {0} .";
```

File: src/Microsoft.Identity.Web/TokenAcquisition.cs [hunk 2]
```diff
@@ -211,93 +211,82 @@ internal class TokenAcquisition : ITokenAcquisition, ITokenAcquisitionInternal
...
+
+        private async Task<string?> GetTokenForWebApiToCallDownstreamApiAsync(
+            IConfidentialClientApplication application,
+            string authority,
+            IEnumerable<string> scopes)
+        {
+            try
+            {
+                // In web API, validatedToken will not be null
+                JwtSecurityToken? validatedToken = CurrentHttpContext.GetTokenUsedToCallWebAPI();
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/TokenAcquisition.cs:266]:
```suggestion
                JwtSecurityToken? validatedToken = CurrentHttpContext?.GetTokenUsedToCallWebAPI();
```

File: src/Microsoft.Identity.Web/TokenAcquisition.cs [hunk 3]
```diff
@@ -588,5 +577,52 @@ private static bool AcceptedTokenVersionMismatch(MsalUiRequiredException msalSev
             // This is subject to change in the future
             return msalSeviceException.Message.Contains(ErrorCodes.B2CPasswordResetErrorCode, StringComparison.InvariantCulture);
         }
+
+        private async Task<ClaimsPrincipal?> GetAuthenticatedUserAsync(ClaimsPrincipal? user)
+        {
+            if (user == null && _httpContextAccessor.HttpContext != null)
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/TokenAcquisition.cs:583]:
```suggestion
            if (user == null && _httpContextAccessor.HttpContext?.User != null)
```

File: src/Microsoft.Identity.Web/Constants/LogMessages.cs (hunk 1 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/Constants/LogMessages.cs:25]: i'm going to call it downstream web api instead, as OBO is not a flow, and it is confusing for a customer. thanks, peter. :)

---
In reply to: [464802916](https://github.com/AzureAD/microsoft-identity-web/pull/383#discussion_r464802916) [](ancestors = 464802916)

File: src/Microsoft.Identity.Web/TokenAcquisition.cs (hunk 3 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/TokenAcquisition.cs:583]: got it. thanks!

---
In reply to: [464811695](https://github.com/AzureAD/microsoft-identity-web/pull/383#discussion_r464811695) [](ancestors = 464811695)

File: src/Microsoft.Identity.Web/TokenAcquisition.cs (hunk 2 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/TokenAcquisition.cs:266]: fixed. thanks

---
In reply to: [464810227](https://github.com/AzureAD/microsoft-identity-web/pull/383#discussion_r464810227) [](ancestors = 464810227)

File: src/Microsoft.Identity.Web/Constants/LogMessages.cs (hunk 1 above)
Comment (pmaytak) [reply src/Microsoft.Identity.Web/Constants/LogMessages.cs]: Yea, downstream API makes sense.

Yea, idk I got the phrase from MS docs: https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow. :)

**Pull request merged**